### PR TITLE
TemplatePart XAML diagnostics

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

1. Adds IsRequired property to the TemplatePart. This property should be set on these parts that control can't work without. For example, `PART_ContentPresenter` is **optional** for Button - it will work without. But `PART_TextPresenter` is necessary for TextBox.
2. Makes TemplatePart inheritable. This one is a bit questionable, as WPF attribute wasn't inheritable on paper. As on practice WPF doesn't duplicate TemplateParts per each inherited type - [TextBoxBase](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.primitives.textboxbase?view=windowsdesktop-8.0) has TemplatePart, while [TextBox](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.textbox?view=windowsdesktop-8.0) doesn't. Also, having it inherited makes compiler job much easier.
3. As a result of previous one - custom control can redefine Type or IsRequired parent TemplatePart. We don't have examples of this in the repo though.
4. **Finally**, this PR adds three diagnostics:
- RequiredTemplatePartMissing = AVLN2205 (enabled as error by default)
- OptionalTemplatePartMissing = AVLN2206 (disabled by default)
- TemplatePartWrongType = AVLN2207 (enabled as error by default)
5. Fixes all AVLN2205 AVLN2207 and some AVLN2206 in the repository.
6. In general I tried to not affect existing behavior much. But some controls have changed their template part requirements. CaptionButtons, for example, now doesn't crash when any button is missing.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

New diagnostic errors might break projects. New errors should be either fixed or disabled with csproj/editorconfig settings.

Part of #13707

cc @robloo @danwalmsley

### CODE REVIEW

**Comment by maxkatz6** [comment]
Having it implemented in XAML compiler makes old Rider YouTrack issue unnecessary. Also, works in VS.

**Comment by robloo** [comment]
A lot of good ideas here! It's also good someone else spent the time going through each TemplatePart. There were so many it's always easy to miss them.

> Adds IsRequired property to the TemplatePart. This property should be set on these parts that control can't work without. For example, PART_ContentPresenter is optional for Button - it will work without. But PART_TextPresenter is necessary for TextBox.

This is a great idea!

> Makes TemplatePart inheritable. This one is a bit questionable, as WPF attribute wasn't inheritable on paper. As on practice WPF doesn't duplicate TemplateParts per each inherited type - [TextBoxBase](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.primitives.textboxbase?view=windowsdesktop-8.0) has TemplatePart, while [TextBox](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.textbox?view=windowsdesktop-8.0) doesn't. Also, having it inherited makes compiler job much easier.

I just followed WPF originally for this. So I have no strong feelings about it and generally think it is a good change. It matches the inheritance of other things related to controls: properties, etc.

 1. However, it is conceivable for a derived control (parent) to completely replace the logic of the child. This means a derived control may not actually require all the template parts of the children. It may have completely different code.
 2. We have to be careful to check all template part attributes to see if there was any duplication... I suspect you already did this?

> As a result of previous one - custom control can redefine Type or IsRequired parent TemplatePart. We don't have examples of this in the repo though.

Yea, since the parent can completely override the child's template part attributes there is less concern. Only in the case of a parent derived control not actually using a child template part would there be concern.

> Finally, this PR adds three diagnostics:

Awesome! This one was sorely needed and you found a good way to do it.

---

Separately, I thought we need to review all `e.NameScope.Find<T>("name")` and `e.NameScope.Get<T>("name")` usages. These really should follow the `IsRequired` property of the attribute. Then I thought they should be automated somehow -- as another analyzer -- but we can't because the name is unknown and there is no way to associate these calls with their corresponding template parts. So I had an idea: We could add the template part `Name` DIRECTLY to the `TemplatePart` attribute. That would allow analyzers to work.

That also logically makes more sense than defining constants all over the place and it really needs to be cleanup up sometime. I wonder if this would solve https://github.com/AvaloniaUI/Avalonia/issues/9093 somehow too. This needs more thought but likely the only way to really make it useful is with a code generator that would automatically generate constant names for the template part attributes.

File: src/Avalonia.Base/Controls/Metadata/TemplatePartAttribute.cs [hunk 1]
```diff
@@ -51,5 +49,10 @@ public TemplatePartAttribute(string name, Type type)
         /// in <see cref="Name"/>.
         /// </summary>
         public Type Type { get; set; }
+
+        /// <summary>
+        /// Gets or sets flag whether template part is mandatory to be present in the template.
```
Comment (robloo) [inline src/Avalonia.Base/Controls/Metadata/TemplatePartAttribute.cs:54]: "Gets or sets a value indicating whether..." is the correct documentation prefix for bool. Some code analyzers will warn if you don't start the summary docs with this.

**Comment by thevortexcloud** [comment]
Thanks for adding this! I was literally asking [for it just last week.](https://github.com/AvaloniaUI/Avalonia/discussions/14131#discussioncomment-8045181)

Is there any plans for a reverse of this too, where a template has parts defined that don't match an attribute?

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlControlTemplatePartsChecker.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                {
+                    dictionary.Add(name, (typeObj as IXamlType, isRequiredObj as bool? == true));
+                }
+            }
+        }
+
+        return dictionary;
+    }
+
+    private class TemplatePartVisitor : Dictionary<string, (IXamlType type, IXamlLineInfo line)>, IXamlAstVisitor
```
[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlControlTemplatePartsChecker.cs:101]: Good one, added a test to ignore nested NestedScopeMetadataNode subtrees. + test.

**Comment by maxkatz6** [comment]
> We have to be careful to check all template part attributes to see if there was any duplication... I suspect you already did this?

I did, but there always a chance to miss something.

**Comment by maxkatz6** [comment]
Updated PR after review. It is ready again.

**Review by kekekeks — APPROVED** [review]

**Pull request merged**